plugins/plugin_utils/consolidate.py

```diff
@@ -0,0 +1,180 @@
+#
+# -*- coding: utf-8 -*-
+# Copyright 2022 Red Hat
+# GNU General Public License v3.0+
+# (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+#
+
+"""
+The consolidate plugin code
+"""
+from __future__ import absolute_import, division, print_function
+
+__metaclass__ = type
+
+from ansible.errors import AnsibleFilterError
+import itertools
+
+
+def _raise_error(filter, msg):
+    """Raise an error message, prepend with filter name
+
+    Args:
+        filter (str): Filter name
+        msg (str): Message specific to filter supplied
+
+    Raises:
+        AnsibleFilterError: AnsibleError with filter name and message
+    """
+    error = "Error when using plugin 'consolidate': '{filter}' reported {msg}".format(
+        filter=filter, msg=msg
+    )
+    raise AnsibleFilterError(error)
+
+
+def fail_on_filter(validator_func):
+    """decorator to fail on supplied filters
+
+    Args:
+        validator_func (func): Function that generates failure messages
+
+    Returns:
+        raw: Value without errors if generated and not failed
+    """
+
+    def update_err(*args, **kwargs):
+
+        res, err = validator_func(*args, **kwargs)
+        if err.get("match_key_err"):
+            _raise_error(
+                "fail_missing_match_key", ", ".join(err["match_key_err"])
+            )
+        if err.get("match_val_err"):
+            _raise_error(
+                "fail_missing_match_value", ", ".join(err["match_val_err"])
+            )
+        if err.get("duplicate_err"):
+            _raise_error("fail_duplicate", ", ".join(err["duplicate_err"]))
+        return res
+
+    return update_err
+
+
+@fail_on_filter
+def check_missing_match_key_duplicate(
+    data_sources, fail_missing_match_key, fail_duplicate
+):
+    """Checks if the match_key specified is present in all the supplied data,
+    also checks for duplicate data accross all the data sources
+
+    Args:
+        data_sources (list): list of dicts as data sources
+        fail_missing_match_key (bool): Fails if match_keys not present in data set
+        fail_duplicate (bool): Fails if duplicate data present in a data
+    Returns:
+        list: list of unique keys based on specified match_keys
+    """
+    results, errors_match_key, errors_duplicate = [], [], []
+    for ds_idx, data_source in enumerate(data_sources):
+        match_key = data_source["match_key"]
```

Comment:
does data_source always has this key?

Reply:
right I must add a `required: True` to them, Thanks